doc/decisions/DR023-Versioning-traits-and-specifications.md

```diff
@@ -0,0 +1,702 @@
+
+# Support all known major versions.
+# Camera
+success_result = TraitsData()
+if (cameraVersion >= 3 and cameraVersion < 4):
+     trait = mediacreation.CameraTrait(success_result)
+     ... # Up to date camera population
+elif (cameraVersion >= 2 and cameraVersion < 1):
+     trait = mediacreation.legacy.CameraTrait_v2(success_result)
+     ... # V2 camera population
+elif (cameraVersion >= 1):
+     trait = mediacreation.legacy.CameraTrait_v1(success_result)
+     ... # V1 camera population
+else:
+  raise RuntimeError()
+
+# Spatial
+if (spatialVersion >= 2 and spatialVersion < 3):
+    trait = mediacreation.SpatialTrait(success_result)
+    ... # Up to date spatial population
+elif (spatialVersion >= 1):
+    trait = mediacreation.legacy.SpatialTrait_v1(success_result)
+     ... # V1 to date spatial population
+else:
+  raise RuntimeError()
+```
+
+### Benefits
+- Maximally flexible, hosts and managers can disregard versioning for
+  traits they don't care about.
+- Obvious from the YAML which version a trait is.
+- Individual traits will only update version when they change, thus they
+  will not have "gaps", they will always go v1->v2->v3.
+- A single trait can be updated in isolation without any effect on other
+  traits in the package.
+- Semver could be used to provide additional runtime compatibility
+  information.
+
+### Drawbacks
+- Version has to be encoded in `traitID` because several methods
+  (including `resolve`) take `traitSet`, which is only a collection of
+  `traitIDs`. There is no `TraitsData` to inspect. This would be a major
+  breaking change as currently conceived.
+- Version support may become fragmented if hosts and managers provide
+  versioned support for only certain traits
+- When packing multiple traits, manager has to branch on each trait as
+  they could be different versions.
+- When extended to specification this _really_ causes problems, as we
+  need to invent a new place to pack specifications version across the
+  wire.
+
+## Version the whole schema.
+This option places a version number is placed at the top of the
+`traits.yml` file or equivalent. This defines the version for all the
+traits within the package.
+
+The manager will still need a mechanism to know which trait-library
+version the host will be expecting. This could be achieved by a
+`HostInterface` method.
+
+```cpp
+//HostInterface implementation
+SchemaVersion HostInterface::schemaVersion(std::string_view library_identifier) {
+  if (library_identifier == "openassetio-mediacreation"){
+    return openassetio_mediacreation::version();
+  }
+}
+```
+### Benefits
+- Version dosen't need to be packed into any `traitId`
+- A traits and specifications library gets a version as a unit, making
+  talking about supported versions easier, and hopefully reducing
+  support fragmentation
+- Manager can branch at a higher level
+
+### Drawbacks
+- Unknown at runtime which traits or specifications have been updated in
```

Comment:
I'm not sure how that can be avoided though, other than automated upgrade/downgrade scripts (which we're trying to avoid due to the development cost).

Regardless how granular the versioning is, the developer still has to adapt their code on a version bump. We need to ensure release notes document the traits that change, of course.

What is the use-case for runtime introspection of individual trait versions?

Reply:
"I'm developing a manger that only deals with locatable content trait and I don't want to have to change my code unless there is a breaking change with that trait ". With package versioning, release notes are the only indication that anything may have changed. There could be hundreds of traits in a package over time.

Runtime introspection comes in the case of "Im a host author publishing my own trait data to a manger that stores and recalls this without concrete understanding" , which I would imagine is pretty common.

I resolve a reference, I can immediately know which trait was stored, and adapt my handling of that data. If you only have the package version, there could be a large number of versions where that specific trait is identical. In many of these situations, and I have to rely on magic knowledge that v2-25 of the package dont need special handling. The host is the only one that can deal with data migration from old to new.

Reply:
☝️ that thing - sorry not been very good at explaining things with typing 🙃

Reply:
Runtime introspection at least allows you to look at the `TraitsData` per-response to learn what data came back. This could easily happen with older assets published a long time ago being mixed with newer assets, where the manager doesn't actively manage that data.

Reply:
> I'm not sure how that can be avoided though, other than automated upgrade/downgrade scripts (which we're trying to avoid due to the development cost).

Random thought - the OTIO metadata (iirc) looks similar to a TraitsData. Could you literally use the same code OTIO uses to define their schemas with up/downgrade to manage traits? You'd just copy the trait ID to an `OTIO_SCHEMA` key?